File: src/packages/sdk-python/src/evalai_sdk/integrations/openai_eval.py
"""openai_chat_eval — local-first regression testing against OpenAI models."""

from __future__ import annotations

import hashlib
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Coroutine, Dict, List, Optional

from evalai_sdk.assertions import expect


@dataclass
class OpenAIChatEvalCase:
    input: str
    expected_output: Optional[str] = None
    test_case_id: Optional[str] = None
    assertions: Optional[List[Dict[str, Any]]] = None


@dataclass
class OpenAIChatEvalCaseResult:
    test_case_id: str
    input: str
    output: str
    expected_output: Optional[str]
    passed: bool
    score: float
    assertions: List[Dict[str, Any]] = field(default_factory=list)
    duration_ms: float = 0
    retries: int = 0
    error: Optional[str] = None


@dataclass
class OpenAIChatEvalResult:
    passed: bool
    name: str
    total: int
    passed_count: int
    failed_count: int
    score: float
    duration_ms: float
    results: List[OpenAIChatEvalCaseResult] = field(default_factory=list)
    retried_cases: int = 0


def _input_hash(text: str) -> str:
    normalized = " ".join(text.strip().split())
    return hashlib.sha256(normalized.encode()).hexdigest()[:12]


async def openai_chat_eval(
    *,
    name: str = "openai-eval",
    model: str = "gpt-4",
    api_key: Optional[str] = None,
    cases: List[OpenAIChatEvalCase],
    system_prompt: Optional[str] = None,
    retries: int = 0,
    temperature: float = 0.0,
    call_fn: Optional[Callable[..., Any]] = None,
    report_to_evalai: bool = False,
    evalai_client: Optional[Any] = None,
    evaluation_id: Optional[int] = None,
) -> OpenAIChatEvalResult:
    """Run a local regression eval suite against an OpenAI-compatible model.

    Requires either ``call_fn`` (a custom async function) or the ``openai`` package.

    Args:
        name: Suite name for reporting.
        model: Model identifier.
        api_key: OpenAI API key (falls back to OPENAI_API_KEY env var).
        cases: List of test cases.
        system_prompt: Optional system prompt prepended to every call.
        retries: Number of retry attempts per case.
        temperature: Sampling temperature.
        call_fn: Optional custom async function(messages) -> str.
        report_to_evalai: If True, upload results to the EvalAI platform.
        evalai_client: AIEvalClient instance (required if report_to_evalai is True).
        evaluation_id: Evaluation ID to associate with (if reporting).

    Returns:
        OpenAIChatEvalResult with per-case results and aggregate score.
    """
    if call_fn is None:
        call_fn = _make_openai_call(model, api_key, system_prompt, temperature)

    results: List[OpenAIChatEvalCaseResult] = []
    suite_start = time.monotonic()
    retried_total = 0

    for case in cases:
        case_id = case.test_case_id or _input_hash(case.input)
        messages = []
        if system_prompt:
            messages.append({"role": "system", "content": system_prompt})
        messages.append({"role": "user", "content": case.input})

        output = ""
        error: Optional[str] = None
        case_retries = 0
        case_start = time.monotonic()

        for attempt in range(1 + retries):
            try:
                output = await call_fn(messages)
                error = None
                break
            except Exception as exc:
                error = str(exc)
                case_retries += 1

        duration = (time.monotonic() - case_start) * 1000

        assertion_results: List[Dict[str, Any]] = []
        case_passed = error is None

        if error is None and case.assertions:
            for a_def in case.assertions:
                a_type = a_def.get("type", "")
                a_value = a_def.get("value")
                exp = expect(output)
                if a_type == "contains" and isinstance(a_value, str):
                    r = exp.to_contain(a_value)
                elif a_type == "not_contains" and isinstance(a_value, str):
                    r = exp.to_not_contain(a_value)
                elif a_type == "equals":
                    r = exp.to_equal(a_value)
                elif a_type == "matches_pattern" and isinstance(a_value, str):
                    r = exp.to_match_pattern(a_value)
                else:
                    continue
                assertion_results.append({"type": a_type, "passed": r.passed, "message": r.message})
                if not r.passed:
                    case_passed = False

        if error is None and case.expected_output is not None and not case.assertions:
            r = expect(output).to_equal(case.expected_output)
            assertion_results.append({"type": "equals", "passed": r.passed, "message": r.message})
            if not r.passed:
                case_passed = False

        if case_retries > 0:
            retried_total += 1

        results.append(OpenAIChatEvalCaseResult(
            test_case_id=case_id,
            input=case.input,
            output=output,
            expected_output=case.expected_output,
            passed=case_passed,
            score=1.0 if case_passed else 0.0,
            assertions=assertion_results,
            duration_ms=duration,
            retries=case_retries,
            error=error,
        ))

    total_duration = (time.monotonic() - suite_start) * 1000
    passed_count = sum(1 for r in results if r.passed)
    score = passed_count / len(results) if results else 0.0

    eval_result = OpenAIChatEvalResult(
        passed=all(r.passed for r in results),
        name=name,
        total=len(results),
        passed_count=passed_count,
        failed_count=len(results) - passed_count,
        score=score,
        duration_ms=total_duration,
        results=results,
        retried_cases=retried_total,
    )

    if report_to_evalai and evalai_client and evaluation_id:
        try:
            from evalai_sdk.types import CreateRunParams
            await evalai_client.evaluations.create_run(evaluation_id, CreateRunParams(
                execution_settings={"score": score, "results_count": len(results)},
            ))
        except Exception:
            pass

    return eval_result


def _make_openai_call(
    model: str,
    api_key: Optional[str],
    system_prompt: Optional[str],
    temperature: float,
) -> Callable[..., Any]:
    """Create an async call function using the OpenAI SDK."""
    async def _call(messages: List[Dict[str, str]]) -> str:
        try:
            import openai
        except ImportError:
            raise ImportError("Install openai: pip install 'evalai-sdk[openai]'")

        import os
        key = api_key or os.environ.get("OPENAI_API_KEY", "")
        client = openai.AsyncOpenAI(api_key=key)
        response = await client.chat.completions.create(
            model=model,
            messages=messages,
            temperature=temperature,
        )
        return response.choices[0].message.content or ""

    return _call

File: src/packages/sdk/src/integrations/openai-eval.ts
/**
 * openAIChatEval — One-function OpenAI chat regression testing
 *
 * Run local regression tests with OpenAI. No EvalAI account required.
 * CI-friendly output. Optional reportToEvalAI in v1.5.
 *
 * @example
 * ```typescript
 * import { openAIChatEval } from '@pauly4010/evalai-sdk';
 *
 * await openAIChatEval({
 *   name: 'chat-regression',
 *   cases: [
 *     { input: 'Hello', expectedOutput: 'greeting' },
 *     { input: '2 + 2 = ?', expectedOutput: '4' }
 *   ]
 * });
 * ```
 */

import { expect } from "../assertions";
import { loadConfig } from "../cli/config";
import type { TestSuiteCaseResult } from "../testing";
import { createTestSuite } from "../testing";
import { sha256Input } from "../utils/input-hash";

const MAX_FAILED_CASES_TO_SHOW = 5;

export interface OpenAIChatEvalCase {
	input: string;
	expectedOutput?: string;
	/** Platform test case ID. When provided, used directly for reportToEvalAI (no input matching). */
	testCaseId?: number;
	assertions?: ((output: string) => import("../assertions").AssertionResult)[];
}

export interface OpenAIChatEvalOptions {
	name: string;
	model?: string;
	apiKey?: string;
	cases: OpenAIChatEvalCase[];
	/** Retry failing cases N times (default: 0). Only failing cases are retried. */
	retries?: number;
	/** v1.5: Upload results to EvalAI platform for an existing evaluation. Requires evaluationId and EVALAI_API_KEY. */
	reportToEvalAI?: boolean;
	/** Evaluation ID (from config or arg). Required when reportToEvalAI is true. */
	evaluationId?: string;
	/** EvalAI API base URL. Default: EVALAI_BASE_URL or http://localhost:3000 */
	baseUrl?: string;
	/** Idempotency key for import (e.g. CI run ID). Prevents duplicate runs on retry. */
	idempotencyKey?: string;
}

export interface OpenAIChatEvalResult {
	passed: number;
	total: number;
	score: number;
	results: TestSuiteCaseResult[];
	durationMs: number;
	/** Case IDs that were retried (flaky recovery) */
	retriedCases?: string[];
}

function getOpenAI(): unknown {
	try {
		const OpenAI = require("openai");
		return OpenAI;
	} catch {
		throw new Error(
			"openai package is required for openAIChatEval. Install with: npm install openai",
		);
	}
}

function createExecutor(
	model: string,
	apiKey: string,
): (input: string) => Promise<string> {
	const OpenAI = getOpenAI() as any;
	const openai = new OpenAI({ apiKey });

	return async (input: string): Promise<string> => {
		const response = await openai.chat.completions.create({
			model,
			messages: [{ role: "user" as const, content: input }],
			temperature: 0.1,
		});
		return response.choices[0]?.message?.content ?? "";
	};
}

function printSummary(result: OpenAIChatEvalResult): void {
	const { passed, total, results } = result;
	const score = total > 0 ? Math.round((passed / total) * 100) : 0;
	const failed = results.filter((r) => !r.passed);

	const status = failed.length === 0 ? "PASS" : "FAIL";
	console.log(`\n${status} ${passed}/${total}  (score: ${score})\n`);

	if (failed.length > 0) {
		const toShow = failed.slice(0, MAX_FAILED_CASES_TO_SHOW);
		const more = failed.length - toShow.length;
		console.log(
			`${failed.length} failing case${failed.length === 1 ? "" : "s"}:`,
		);
		for (const r of toShow) {
			const expected = r.expected ?? "(no expected)";
			console.log(`- "${r.input}" → expected: ${expected}`);
		}
		if (more > 0) {
			console.log(`+ ${more} more`);
		}
		console.log("\nGate this in CI:");
		console.log("  npx -y @pauly4010/evalai-sdk@^1 init");
	} else {
		console.log("Tip: Want dashboards and history?");
		console.log("Set EVALAI_API_KEY and connect this to the platform.");
	}
}

/**
 * Run OpenAI chat regression tests locally.
 * No EvalAI account required. Returns score and prints CI-friendly summary.
 */
export async function openAIChatEval(
	options: OpenAIChatEvalOptions,
): Promise<OpenAIChatEvalResult> {
	const { name, model = "gpt-4o-mini", apiKey, cases, retries = 0 } = options;

	const resolvedApiKey =
		apiKey ?? (typeof process !== "undefined" && process.env?.OPENAI_API_KEY);
	if (!resolvedApiKey) {
		throw new Error(
			"OPENAI_API_KEY is required. Set it in the environment or pass apiKey to openAIChatEval.",
		);
	}

	const executor = createExecutor(model, resolvedApiKey);

	const suiteCases = cases.map((c) => {
		const assertions = c.assertions
			? [...c.assertions]
			: c.expectedOutput
				? [
						(output: string) =>
							expect(output).toContainKeywords(
								c.expectedOutput!.split(/\s+/).filter(Boolean),
							),
					]
				: undefined;
		return {
			input: c.input,
			expected: c.expectedOutput,
			assertions,
		};
	});

	const suite = createTestSuite(name, {
		cases: suiteCases,
		executor,
		parallel: true,
		retries,
	});

	const result = await suite.run();

	const score =
		result.total > 0 ? Math.round((result.passed / result.total) * 100) : 0;
	const evalResult: OpenAIChatEvalResult = {
		passed: result.passed,
		total: result.total,
		score,
		results: result.results,
		durationMs: result.durationMs,
		...(result.retriedCases &&
			result.retriedCases.length > 0 && { retriedCases: result.retriedCases }),
	};

	printSummary(evalResult);

	// v1.5: Optional report to EvalAI platform
	if (options.reportToEvalAI) {
		const config =
			typeof process !== "undefined" && process.cwd
				? loadConfig(process.cwd())
				: null;
		const evalId = options.evaluationId || config?.evaluationId;
		if (!evalId || String(evalId).trim() === "") {
			console.log("Run evalai init and set evaluationId to upload results.");
			return evalResult;
		}
		const evalaiKey =
			(typeof process !== "undefined" && process.env?.EVALAI_API_KEY) || "";
		if (!evalaiKey) {
			console.log("Set EVALAI_API_KEY to upload results.");
			return evalResult;
		}
		const baseUrl =
			options.baseUrl ||
			config?.baseUrl ||
			(typeof process !== "undefined" && process.env?.EVALAI_BASE_URL) ||
			"http://localhost:3000";
		const url = String(baseUrl).replace(/\/$/, "");

		try {
			// Resolve testCaseId for each result: explicit testCaseId in cases, or match by inputHash
			const importResults: Array<{
				testCaseId: number;
				status: "passed" | "failed";
				output: string;
				latencyMs?: number;
			}> = [];

			const hasExplicitIds = cases.some((c) => c.testCaseId != null);
			if (hasExplicitIds) {
				// Use testCaseId from cases (same order as results)
				for (let i = 0; i < result.results.length; i++) {
					const tcId = cases[i]?.testCaseId;
					if (tcId == null) {
						console.log(
							"reportToEvalAI: All cases must have testCaseId when unknown has it.",
						);
						return evalResult;
					}
					importResults.push({
						testCaseId: tcId,
						status: result.results[i].passed ? "passed" : "failed",
						output: result.results[i].actual ?? "",
						latencyMs: result.results[i].durationMs,
					});
				}
			} else {
				// Match by inputHash (same canonicalization as platform)
				const tcRes = await fetch(
					`${url}/api/evaluations/${evalId}/test-cases?limit=500`,
					{
						headers: { Authorization: `Bearer ${evalaiKey}` },
					},
				);
				if (!tcRes.ok) {
					console.log(
						"Could not fetch test cases. Check evaluationId and EVALAI_API_KEY.",
					);
					return evalResult;
				}
				const platformCases = (await tcRes.json()) as Array<{
					id: number;
					input: string;
				}>;
				const hashToIds = new Map<string, number[]>();
				for (const tc of platformCases) {
					const input = tc.input ?? "";
					if (!input.trim()) continue;
					const hash = sha256Input(input);
					const existing = hashToIds.get(hash) ?? [];
					existing.push(tc.id);
					hashToIds.set(hash, existing);
				}

				for (const r of result.results) {
					const hash = sha256Input(r.input ?? "");
					const ids = hashToIds.get(hash);
					if (ids == null || ids.length === 0) {
						console.log(
							`No platform test case matches input: "${(r.input ?? "").slice(0, 50)}…"`,
						);
						return evalResult;
					}
					if (ids.length > 1) {
						console.log(
							`Multiple platform test cases share the same input (hash collision). Use testCaseId in cases.`,
						);
						return evalResult;
					}
					importResults.push({
						testCaseId: ids[0],
						status: r.passed ? "passed" : "failed",
						output: r.actual ?? "",
						latencyMs: r.durationMs,
					});
				}
			}

			if (importResults.length !== result.results.length) {
				console.log("Could not match all results to platform test cases.");
				return evalResult;
			}

			const sdkVersion = "1.4.1";
			const headers: Record<string, string> = {
				"Content-Type": "application/json",
				Authorization: `Bearer ${evalaiKey}`,
			};
			if (options.idempotencyKey) {
				headers["Idempotency-Key"] = options.idempotencyKey;
			}
			const importRes = await fetch(
				`${url}/api/evaluations/${evalId}/runs/import`,
				{
					method: "POST",
					headers,
					body: JSON.stringify({
						environment: "dev",
						results: importResults,
						importClientVersion: sdkVersion,
					}),
				},
			);

			if (!importRes.ok) {
				const body = await importRes.text();
				console.log(`Upload failed: ${importRes.status} — ${body}`);
				return evalResult;
			}

			const importData = (await importRes.json()) as { dashboardUrl?: string };
			if (importData.dashboardUrl) {
				console.log(`Dashboard: ${importData.dashboardUrl}`);
			}
		} catch (err) {
			console.log(
				"Upload failed:",
				err instanceof Error ? err.message : String(err),
			);
		}
	}

	return evalResult;
}
